File: src/PrintDiamond/PrintDiamond.cs
using System.Collections.Generic;
using System.Linq;

namespace PrintDiamond;

static class PrintDiamond
{
    internal const char Space = '-';
    internal const char Newline = '\n';

    internal static string Print(char upToChar)
    {
        var n = upToChar - 'a' + 1;

        return
            Enumerable.Range(0, n)
                .Select(index =>
                    BuildLine(index, n))
                .Select(SemiDuplicate)
                .SemiDuplicate()
                .Joined();
    }

    private static IEnumerable<char> BuildLine(int index, int n)
    {
        var leadingSpaces = Spaces(n - index - 1);
        var theChar = (char)('a' + index);
        var trailingSpaces = Spaces(index);
        var row = leadingSpaces.Append(theChar).Concat(trailingSpaces);

        return row;
    }

    private static IEnumerable<char> Spaces(int numberOfSpaces) =>
        Enumerable.Repeat(Space, numberOfSpaces);

    private static string Joined(this IEnumerable<IEnumerable<char>> lines) =>
        string.Join(Newline, lines.Select(c => new string(c.ToArray())));

    private static IEnumerable<T> SemiDuplicate<T>(this IEnumerable<T> xs) =>
        xs.Concat(xs.Reverse().Skip(1));
}

File: src/PrintDiamondFSharp/PrintDiamond.fs
module PrintDiamondFSharp.PrintDiamond

let space = '-'
let newLine = '\n'

let spaces numberOfSpaces =
    String.replicate numberOfSpaces (string space)

let buildLine index n =
    let leadingSpaces = spaces (n - index - 1)
    let theChar = char (int 'a' + index)
    let trailingSpaces = spaces index
    leadingSpaces + string theChar + trailingSpaces

let semiDuplicate xs =
    Seq.append xs (Seq.rev xs |> Seq.skip 1)

let print (upToChar: char) =
    let length = int upToChar - int 'a' + 1

    let semiDuplicate =
        [ 0 .. length - 1 ]
        |> Seq.map (fun index -> buildLine index length)
        |> Seq.map semiDuplicate
        |> Seq.map System.String.Concat
        |> semiDuplicate

    semiDuplicate |> fun s -> System.String.Join(newLine, s)
